### PR TITLE
clean up connections always

### PR DESCRIPTION
**Pull request by ramaraochavali**
I have seen once Istiod has stale proxies listed in `/connections` debug endpoint. This the rough sequence
- New version Istiod pods came up
- Service pods connected to Istiod
- Service went through rolling restart immediately while Istiod is still serving config to the connected old pods
- Old service pods still show up in `/connections`

Not able to reproduce this consistently. 

This PR attempts to fix it by moving the defer function to the top so that it gets executed always.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pilot/pkg/xds/ads.go [hunk 1]
```diff
@@ -508,16 +504,15 @@ func (s *DiscoveryServer) initConnection(node *core.Node, con *Connection) error
 		con.proxy.VerifiedIdentity = id
 	}
 
-	// Register the connection; this allows pushes to be triggered for the proxy. Note: the timing of
-	// this an initProxyState is important. While registering for pushes *after* initialization is
```
Comment (howardjohn) [inline pilot/pkg/xds/ads.go:512]: seems like the "and initProxyState" is important part of the comment, despite a typo

**Review by howardjohn — COMMENTED** [review]
I am not sure I understand how this fixes  things. Why do we need to removeCon if we never hit initConnection?

**Comment by ramaraochavali** [comment]
Reading the code, this can only happen if initConnection adds the connection to list of connections here https://github.com/istio/istio/blob/master/pilot/pkg/xds/ads.go#L517 and initProxyState returns error https://github.com/istio/istio/blob/master/pilot/pkg/xds/ads.go#L520 here and  we return early here https://github.com/istio/istio/blob/master/pilot/pkg/xds/ads.go#L183. Otherwise I could not think of how this can happen. That is why I moved the removeCon to be executed always even in case of error.

But as I said, I could not reproduce this - I have seen it happen once for sure. So this fix based on code path.

**Review by howardjohn — COMMENTED** [review]
maybe we can just cleanup in initConnection if we have an error in initConnection. The contract of initConnection should be that it has no side effects when there is an error?

**Comment by ramaraochavali** [comment]
> maybe we can just cleanup in initConnection if we have an error in initConnection. The contract of initConnection should be that it has no side effects when there is an error?

That makes total sense. Let me change that

**Comment by ramaraochavali** [comment]
@howardjohn changed. PTAL

File: pilot/pkg/xds/ads.go [hunk 2]
```diff
@@ -508,29 +503,43 @@ func (s *DiscoveryServer) initConnection(node *core.Node, con *Connection) error
...
 		s.StatusGen.OnConnect(con)
 	}
 	return nil
 }
 
+func (s *DiscoveryServer) closeConnection(con *Connection) {
+	if con.ConID == "" {
+		return
+	}
+	log.Infof("ADS: connection closed for node:%s", con.ConID)
```
Comment (howardjohn) [inline pilot/pkg/xds/ads.go:532]: I think we already have this from `... terminated ...` logs in receive

File: pilot/pkg/xds/ads.go [hunk 3]
```diff
@@ -508,29 +503,43 @@ func (s *DiscoveryServer) initConnection(node *core.Node, con *Connection) error
...
+	// push context, leading the proxy to have a stale state until the next full push.
 	s.addCon(con.ConID, con)
+	// Register that initialization is complete. This triggers to calls that it is safe to access the
+	// proxy
+	defer close(con.initialized)
 
 	// Complete full initialization of the proxy
-	if err := s.initProxyState(node, con); err != nil {
+	if err := s.initializeProxy(node, con); err != nil {
+		s.closeConnection(con)
```
Comment (howardjohn) [inline pilot/pkg/xds/ads.go:518]: if this happens we will do removeCon 2x, which will increment a metric and give a confusing error of "ADS: Removing connection for non-existing node".

File: pilot/pkg/xds/ads.go (hunk 2 above)
Comment (ramaraochavali) [reply pilot/pkg/xds/ads.go]: removed

File: pilot/pkg/xds/ads.go (hunk 3 above)
Comment (ramaraochavali) [reply pilot/pkg/xds/ads.go:518]: good point. Moved defer to later

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/xds/delta.go [hunk 4]
```diff
@@ -215,6 +215,9 @@ func (s *DiscoveryServer) receiveDelta(con *Connection, reqChannel chan *discove
 			close(con.initialized)
 		}
 	}()
+	defer func() {
+		s.closeConnection(con)
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply pilot/pkg/xds/delta.go:219]: I will fix that.

**Pull request merged**